src/util/sandbox/fd_sandbox.c

```diff
@@ -1,188 +1,138 @@
-    }
+  for( ulong i=0UL; i<allowed_file_descriptor_cnt; i++ ) {
+    if( allowed_file_descriptor[ i ]<0 || allowed_file_descriptor[ i ]==INT_MAX )
+      FD_LOG_ERR(( "allowed_file_descriptors contains invalid file descriptor %d", allowed_file_descriptor[ i ] ));
   }
 
-  for( ulong i=0; i<allow_fds_cnt; i++ ) {
-    if( FD_UNLIKELY( !seen_fds[ i ] ) ) {
-      FD_LOG_ERR(( "allowed file descriptor %d not present", allow_fds[ i ] ));
+  for( ulong i=0UL; i<allowed_file_descriptor_cnt; i++ ) {
+    for( ulong j=0UL; j<allowed_file_descriptor_cnt; j++ ) {
+      if( i==j ) continue;
+      if( allowed_file_descriptor[ i ]==allowed_file_descriptor[ j ] )
+        FD_LOG_ERR(( "allowed_file_descriptor contains duplicate entry %d", allowed_file_descriptor[ i ] ));
     }
   }
 
-  FD_TESTV( !closedir( dir ) );
-}
+  int dirfd = open( "/proc/self/fd", O_RDONLY | O_DIRECTORY );
+  if( dirfd<0 ) FD_LOG_ERR(( "open(/proc/self/fd) failed (%i-%s)", errno, fd_io_strerror( errno ) ));
+
+  for(;;) {
+    uchar buf[ 4096 ];
+    long dents_bytes = syscall( SYS_getdents64, dirfd, buf, sizeof( buf ) );
+    if( !dents_bytes ) break;
+    else if( -1L==dents_bytes ) FD_LOG_ERR(( "getdents64() failed (%i-%s)", errno, fd_io_strerror( errno ) ));
+
+    ulong offset = 0UL;
+    while( offset<(ulong)dents_bytes ) {
+      struct dirent64 const * dent = (struct dirent64 const *)(buf + offset);
+      if( !strcmp( dent->d_name, "." ) || !strcmp( dent->d_name, ".." ) ) {
+        offset += dent->d_reclen;
+        continue;
+      }
 
-static void
-install_seccomp( ushort               seccomp_filter_cnt,
-                 struct sock_filter * seccomp_filter ) {
-  struct sock_fprog program = {
-    .len    = seccomp_filter_cnt,
-    .filter = seccomp_filter,
-  };
-  FD_TESTV( 0 == syscall( SYS_seccomp, SECCOMP_SET_MODE_FILTER, 0, &program ) );
-}
+      char * end;
+      long _fd = strtol( dent->d_name, &end, 10 );
+      if( *end != '\0' ) FD_LOG_ERR(( "/proc/self/pid has unrecognized entry name %s", dent->d_name ));
+      if( _fd>=INT_MAX ) FD_LOG_ERR(( "/proc/self/pid has file descriptor number %ld which is too large", _fd ));
+      int fd = (int)_fd;
 
-static void
-drop_capabilities( void ) {
-  FD_TESTV( 0 == prctl (
-    PR_SET_SECUREBITS,
-    SECBIT_KEEP_CAPS_LOCKED | SECBIT_NO_SETUID_FIXUP |
-      SECBIT_NO_SETUID_FIXUP_LOCKED | SECBIT_NOROOT |
-      SECBIT_NOROOT_LOCKED | SECBIT_NO_CAP_AMBIENT_RAISE |
-      SECBIT_NO_CAP_AMBIENT_RAISE_LOCKED ) );
+      if( fd==dirfd ) {
+        offset += dent->d_reclen;
+        continue;
+      }
 
-  struct __user_cap_header_struct hdr = { _LINUX_CAPABILITY_VERSION_3, 0 };
-  struct __user_cap_data_struct   data[2] = { { 0 } };
-  FD_TESTV( 0 == syscall( SYS_capset, &hdr, data ) );
+      int found = 0;
+      for( ulong i=0UL; i<allowed_file_descriptor_cnt; i++ ) {
+        if( fd==allowed_file_descriptor[ i ] ) {
+          seen_fds[ i ] = 1;
```

Comment:
It feels like you should check that `seen_fds[ i ] == 0` here, but I don't really think that can happen. In a general directory, it could happen because `strtol` discards whitespace, but filenames in `/proc/self/pid` can't have whitespace.